File: 백준/Gold/22859. HTML 파싱/HTML 파싱.js
const input = require("fs").readFileSync(process.platform==='linux'?0:'input.txt').toString().trim();
const N = input.length - 7;
let answer = "";

let i = 6;
while (i < N) {
  const start = input.indexOf("<", i);
  const end = input.indexOf(">", i);
  const tag = input.slice(start + 1, end);
  i = end + 1;
  
  if (tag[0] === "/") continue;

  // p tag
  if (tag === "p") {
    let content = "";
    while (i < N) {
      const start = input.indexOf("<", i);
      const end = input.indexOf(">", i);
      const tag = input.slice(start + 1, end);
      content += input.slice(i, start);
      i = end + 1;

      if (tag !== "/p") continue;

      content = content.trim().split(" ").filter(x => x).join(" ");
      answer += content + "\n";
      break;
    }
  } 
  // div tag
  else {
    answer += "title : " + tag.slice(11, tag.length - 1) + "\n";
  }
}

console.log(answer);

File: 백준/Gold/22859. HTML 파싱/HTML 파싱.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br  = new BufferedReader(new InputStreamReader(System.in));
        String[] input = br.readLine().split("");
        StringBuilder sb = new StringBuilder();

        int i = 0;
        while (i < input.length) {
            // tag appears
            if (input[i].equals("<")) {
                // tag
                String tag = input[++i];
                while (!input[++i].equals(">")) {
                    tag += input[i];
                }

                // opening tag
                if (tag.charAt(0) != '/') {
                    // div
                    if (tag.length() > 3 && tag.substring(0, 3).equals("div")) {
                        String title = tag.substring(11, tag.length()-1);
                        sb.append("title : ").append(title).append("\n");
                    // p
                    } else if (tag.equals("p")) {
                        i++;
                        String content = "";
                        while (i < input.length) {
                            // subTag appears
                            if (input[i].equals("<")) {
                                String subTag = input[++i];
                                while (!input[++i].equals(">")) {
                                    subTag += input[i];
                                }
                                // closing pTag
                                if (subTag.equals("/p")) {
                                    content = content.trim();
                                    // trim white space
                                    String tmp = "";
                                    int j = 0;
                                    while (j < content.length()) {
                                        if (content.charAt(j) == ' ') {
                                            while (content.charAt(j) == ' ') {
                                                j++;
                                            }
                                            tmp += " ";
                                            continue;
                                        }
                                        tmp += content.charAt(j++);
                                    }
                                    sb.append(tmp).append("\n");
                                    break;
                                }
                            // content
                            } else {
                                content += input[i];
                            }
                            i++;
                        }
                    }
                }
            }

            i++;
        }

        System.out.println(sb.toString());
    }
}